### DISCUSSION TITLE
Best Practices for create plugins using json_schema response format with auto invoke function calling

### DISCUSSION THREAD

**Discussion by 4jonesy** [microsoft/semantic-kernel#9475] [General]
I am wondering if there are any best practices for using the json_schema response format in SK plugins using auto invoke function calling.

For example, I currently have a plugin created for a product which contains a number of KernelFunction defined methods for getting lists of data, details of a specific item, creating and updating items, etc...  Each of these methods return a different json response/schema.

Before, I was using the json mode as opposed to json_schema response format.  With that, I could return the json schema from my method and then look back in the chat history to grab the last tool response to get the actual json response returned from the kernel function method and return it to the ui for data rendering.

However, when using json_schame resposne format, it appears that there is only one single json_schema that be applied across all calls to the chatcompletionservice.  See code snippet below where I am using a custom model class to define my json_schema:

```
     OpenAIPromptExecutionSettings openAIPromptExecutionSettings = new OpenAIPromptExecutionSettings
     {
            ToolCallBehavior = ToolCallBehavior.AutoInvokeKernelFunctions,
            ResponseFormat = typeof(CustomModelClass)
     };

     var result = await chatCompletionService.GetChatMessageContentAsync(
            chatHistory,
            executionSettings: openAIPromptExecutionSettings,
     kernel: _kernel);
```

I feel the way I have it currently is kind of a 'improper/hacky' way to do it.  Using the json_schema response format, is there a way I can return a different json schema depending on what sk function is auto invoked within the same chat instance?

or would I need to create some larger json schema that would incorporate each sk methods response?  That seems kind of bulky to me.

Or are there any other suggestions on what I can do?

**Comment by moonbox3** [comment]
Hi @4jonesy, the way that OpenAI is supporting the `json_schema` response format right now is that they only allow one `strict` schema, just as you pointed out. You would need to find a way to be able to provide this single response format schema, which for your use-case, may be larger than desired. This is beta functionality from OpenAI and we'll keep an eye on any changes/enhancements if/when they appear.
